autoscaler/nanny/kubernetes_client.go

```diff
@@ -0,0 +1,161 @@
+*/
+
+package nanny
+
+import (
+	"fmt"
+	"log"
+
+	api "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/resource"
+	apiv1 "k8s.io/kubernetes/pkg/api/v1"
+	client "k8s.io/kubernetes/pkg/client/clientset_generated/release_1_3"
+	"k8s.io/kubernetes/pkg/client/restclient"
+)
+
+// KubernetesClient - Wraps all Kubernetes API client functionality
+type KubernetesClient struct {
+	namespace  string
+	rc         string
+	rs         string
+	deployment string
+	clientset  *client.Clientset
+}
+
+// FetchConfigMap - fetch the requested key from the configmap
+func (k *KubernetesClient) FetchConfigMap(namespace, configmap, key string) (data string, version string, err error) {
+	cm, err := k.clientset.CoreClient.ConfigMaps(k.namespace).Get(configmap)
+	if err != nil {
+		return "", "", err
+	}
+	return cm.Data[key], cm.ObjectMeta.ResourceVersion, nil
+}
+
+// CountNodes Count schedulable nodes and cores in our cluster
+func (k *KubernetesClient) CountNodes() (totalNodes, schedulableNodes, totalCores, schedulableCores int32, err error) {
+	opt := api.ListOptions{Watch: false}
+
+	nodes, err := k.clientset.CoreClient.Nodes().List(opt)
+	if err != nil {
+		log.Println(err)
+		return 0, 0, 0, 0, err
+	}
+	totalNodes = int32(len(nodes.Items))
+	var tc resource.Quantity
+	var sc resource.Quantity
+	for _, node := range nodes.Items {
+		tc.Add(node.Status.Capacity[apiv1.ResourceCPU])
+		if !node.Spec.Unschedulable {
+			schedulableNodes++
+			sc.Add(node.Status.Capacity[apiv1.ResourceCPU])
+		}
+	}
+
+	tcInt64, tcOk := tc.AsInt64()
+	scInt64, scOk := sc.AsInt64()
+	if !tcOk || !scOk {
+		log.Println("Unable to compute integer values of schedulable cores in the cluster")
+		return 0, 0, 0, 0, fmt.Errorf("Unable to compute number of cores in cluster")
+	}
+	return totalNodes, schedulableNodes, int32(tcInt64), int32(scInt64), nil
+}
+
+// PodReplicas Get number of replicas configured in the parent RC/Deployment/RS, in that order
+func (k *KubernetesClient) PodReplicas() (int32, error) {
+	if len(k.rc) > 0 {
+		rc, err := k.clientset.CoreClient.ReplicationControllers(k.namespace).Get(k.rc)
+		if err != nil {
+			log.Printf("Failed to fetch RC %s/%s (%s)\n", k.namespace, k.rc, err)
+			return 0, err
+		}
+		return int32(*rc.Spec.Replicas), nil
+	} else if len(k.deployment) > 0 {
+		deployment, err := k.clientset.ExtensionsClient.Deployments(k.namespace).Get(k.deployment)
+		if err != nil {
+			return 0, err
+		}
+		return int32(*deployment.Spec.Replicas), nil
+	} else if len(k.rs) > 0 {
+		rs, err := k.clientset.ExtensionsClient.ReplicaSets(k.namespace).Get(k.rs)
+		if err != nil {
+			return 0, err
+		}
+		return int32(*rs.Spec.Replicas), nil
+	}
+	return 0, fmt.Errorf("Invalid RC/RS/Deployment configured")
+}
+
+// UpdateReplicas Update the number of replicas in the controller
+func (k *KubernetesClient) UpdateReplicas(replicas int32) error {
+	if replicas == 0 {
+		log.Fatalf("Cannot update to 0 replicas")
+	}
+	if len(k.rc) > 0 {
```

Comment:
I can't find your other comment "kubectl makes this look easy" (thanks, github), so I just want to respond with "this comment" :)